### ISSUE TITLE
PROJ not following the CMake convention of BUILD_SHARED_LIBS

### ISSUE THREAD

**Issue by ArtificialRaccoon** [OSGeo/PROJ#1858]
_Directed by hobu on glitter to raise a ticket_

#### Example of problem
Using the PROJ6.3.0 source release, I am unable to build proj.dll (everything else builds fine).  According to what I have read, I should be able to add the following to the cmakelists.txt to produce a DLL:

set(BUILD_SHARED_LIBS ON)

However, this has no effect.  PROJ successfully builds the .LIB, and the executable files; but the .DLL is not created.  This is the command I am using from the VS2019 Command Prompt to start my build:

```
cmake -DCMAKE_GENERATOR_PLATFORM=x64 -DSQLITE3_INCLUDE_DIR=....\ThirdParty\sqlite\include -DSQLITE3_LIBRARY=....\ThirdParty\sqlite\Release\lib\sqlite3.lib -DEXE_SQLITE3=....\ThirdParty\sqlite\Release\bin\sqlite3.exe -DCMAKE_INSTALL_PREFIX=Release ..

cmake --build . --config RELEASE --target INSTALL
```

#### Problem description

For us, this is an issue, as we use the DLL when building a custom C# wrapper for the GDAL.  We have found a way to generate the DLL, but it involves modifying the lib_proj.cmake file:

```
if(WIN32)
  set(BUILD_LIBPROJ_SHARED_DEFAULT ON)
endif()
option(BUILD_LIBPROJ_SHARED_DEFAULT
  "Build libproj library shared." ${BUILD_LIBPROJ_SHARED_DEFAULT})
if(BUILD_LIBPROJ_SHARED_DEFAULT)
  set(PROJ_LIBRARY_TYPE SHARED)
else()
  set(PROJ_LIBRARY_TYPE STATIC)
endif()
```

However, as one would expect, this breaks the build in other ways. The DLL will now build, but the LIB and EXE's will no longer build. (I should note, I'm not a C/C++ developer).

We are hoping to have a reproducible build process if possible.

#### Expected Output
A build of the DLL, LIB, and EXEs.

#### Environment Information
 - PROJ version (`proj`)
PROJ 6.3.0 (https://github.com/OSGeo/PROJ/releases/tag/6.3.0)

 - Operation System Information
Windows 10, Visual Studio 2019

#### Installation method
Downloaded the source release from (https://github.com/OSGeo/PROJ/releases/tag/6.3.0), un-archived it, and attempted to build.   See "Example of problem" above for the exact calls to cmake I'm doing to start the build after extracting the archive.

**Comment by rouault** [comment]
Indeed, PROJ uses a non standard CMake convention. To build the DLL, you need to pass -DBUILD_LIBPROJ_SHARED=ON. This is what PROJ uses for appveyor.yml

**Comment by hobu** [comment]
My point is we shouldn't be doing anything non-standard in regards to BUILD_SHARED_LIBS. Our library isn't special, and this is an expected convention by CMake users

**Comment by ArtificialRaccoon** [comment]
Just popping back in:
I performed the build from a fresh copy of source, using the DBUILD_LIBPROJ_SHARED paramter.  The build was successful, and produced the LIB, DLL, and EXEs.  I can use this to create a reproducable build; thank you.

My two cents:  I am not opposed to non-standard items.  However, I had no idea this existed.  Following the Github Wiki page to it's logical conclusion, I ended up at the install documents (https://proj.org/install.html).  This parameter isn't mentioned in the Visual Studio / CMake section.  

Admittedly, my google-fu could also be lacking, but I didn't find any mention of  non-standard cmake parameters (at least in an obvious location).  After googling "DBUILD_LIBPROJ_SHARED", I see many references to it; though mostly in closed github issues.

Perhaps the easiest solution is a small addition to the documents which mention the DBUILD_LIBPROJ_SHARED parameter and what is does?

**Comment by kbevers** [comment]
> My point is we shouldn't be doing anything non-standard in regards to BUILD_SHARED_LIBS. Our library isn't special, and this is an expected convention by CMake users

I aggree that we should try to conform to standard use patterns as much as possible. Would you be able update the CMake configuration in time for the 7.0 release @hobu?

**Comment by hobu** [comment]
What is the latest minimum CMake we can get away with requiring? PDAL currently requires 3.5, but we'd like to bump that even higher if we can. 

https://kubasejdak.com/modern-cmake-is-like-inheritance

**Comment by kbevers** [comment]
> What is the latest minimum CMake we can get away with requiring?

@mwtoews suggested that 3.9 would be the best option at the moment:  https://lists.osgeo.org/pipermail/proj/2020-January/009245.html

**Comment by hobu** [comment]
Because we have both autoconf and cmake, I would suggest we bump the version number higher than 3.9 – maybe something like 3.13, which provides better target stuff and fixes a few significant bugs.

**Comment by kbevers** [comment]
As long as you can stay within the rules of [RFC3](https://proj.org/community/rfc/rfc-3.html) I don't care which version of CMake is used.

**Comment by mwtoews** [comment]
[CMake 3.13.0](https://blog.kitware.com/cmake-3-13-0-available-for-download/) was released  November 20, 2018, so it a bit too new for RFC3's two years old threshold. Given this constraint, the maximum version is [CMake 3.10.0](https://blog.kitware.com/cmake-3-10-0-available-for-download/), released November 20, 2017, but I don't see anything too compelling to recommend it to the PSC as the next minimum version.

**Comment by mwtoews** [comment]
@kbevers is this getting too late to try and fix for v7.0? I'm not entirely sure how long it will take to set-up and test on a Windows PC.

**Comment by kbevers** [comment]
>  is this getting too late to try and fix for v7.0?

Pushing the limit I'd say but if you can get it ready for 7.0.0RC2 I think it is okay. I plan on releasing RC2 on Monday the 24th (although it may end up being the 25th as I have a cramped schedule on Monday).
